tools/mtypes/main.go

```diff
@@ -0,0 +1,217 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+type stats struct {
+	families, series, buckets, objectives int
+
+	// adjustedSeries represents series that would result in "series" in Prometheus data model
+	// (includes _bucket, _count, _sum, _quantile).
+	adjustedSeries int
+}
+
+var metricType_NATIVE_HISTOGRAM dto.MetricType = 999
+
+func main() {
+	resource := flag.String("resource", "", "Path or URL to the resource (file, <url>/metrics) containing Prometheus metric format.")
+	avalancheFlagsForTotal := flag.Int("avalanche-flags-for-adjusted-series", 0, "If more than zero, it additionally prints flags for the avalanche 0.6.0 command line to generate metrics for the similar type distribution; to get the total number of adjusted series to the given value.")
+	flag.Parse()
+
+	var input io.Reader = os.Stdin
+	if *resource != "" {
+		switch {
+		case strings.HasPrefix(*resource, "https://"), strings.HasPrefix(*resource, "http://"):
+			if _, err := url.Parse(*resource); err != nil {
+				log.Fatalf("error parsing HTTP URL to the resource %v; got %v", *resource, err)
+			}
+			resp, err := http.Get(*resource)
+			if err != nil {
+				log.Fatalf("http get against %v failed", err)
+			}
+			defer resp.Body.Close()
+			input = resp.Body
+		default:
+			// Open the input file.
+			file, err := os.Open(*resource)
+			if err != nil {
+				log.Fatalf("Error opening file: %v", err)
+			}
+			defer file.Close()
+			input = file
+		}
+	}
+	statistics, err := calculateTargetStatistics(input)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var total stats
+	for _, s := range statistics {
+		total.families += s.families
+		total.series += s.series
+		total.adjustedSeries += s.adjustedSeries
+	}
+
+	writeStatistics(os.Stdout, total, statistics)
+
+	if *avalancheFlagsForTotal > 0 {
+		adjustedGoal := float64(*avalancheFlagsForTotal)
+		fmt.Println()
+		fmt.Println("Avalanche flags for the similar distribution to get to the adjusted series goal of:", adjustedGoal)
+
+		adjustedGoal = adjustedGoal / 10.0 // Assuming --series-count=10
+		adjustedSum := 0
+		for _, mtype := range allTypes {
```

Comment:
There is so much math here, I keep getting confused.

Maybe we can add a comment here to explain the algorithm? Or rename some of the vars to explain?

- `adjustedGoal` is tracking the # of series of that metric type that avalanche will have to create
- `adjustedSum` is tracking the total sum of series so far (at the end hopefully adjustedSum ~= adjustedGoal)
- `adjustedSeries` is tracking (per metric type), how many unique series of that metric type avalanche needs to create according to the ratio we got from our input example

If all this makes sense, then I think I have my head wrapped around all the math. Correct me if I got something wrong

Reply:
Yup, we can add more comments, will add later. There is definitely room to rename metrics for clarity. Maybe `adjustedSeries` should be `adjustedSeriesForType`?

Reply:
Otherwise your explanation is 100% right, nice!